Review Проектная работа. Redux-хранилище "Stellar Burger"

tryproxy opened the pull request:
1. Task: Реализовать маршруты, авторизацию и глобальное состояние приложения.
2. Screenshot: 
![image](https://github.com/user-attachments/assets/5038ca75-1438-49ae-8479-f27df662260c)
3. Done 22.07.2024

Реализовано:
- [x] подключить роутер и реализовать маршруты;
- [X] добавить глобальное состояние и подключить к страницам;
- [X] реализовать механизм авторизации и защитить личный кабинет пользователя;

tryproxy committed:
feat: add initial routes to app

tryproxy committed:
feat: add ingredient slice

tryproxy committed:
feat: add NavLinks to header;
feat: add feedSlice reducer;

tryproxy committed:
fix: Feed fetch;
feat: add modal order Feed;

tryproxy committed:
feat: add burgerConstructorSlice;

tryproxy committed:
fix: unique ingredient ID when added to burgerConstructor

tryproxy committed:
fix: removeitem by unique ID from burgerConstructor

tryproxy committed:
feat: add profile slice w/ registration async action

tryproxy committed:
feat: add login/logout functionality

tryproxy committed:
feat: add order slice, add order burger functionality

tryproxy committed:
feat: add update user credentials; add orders user history to profile;
add redirect in constructor burger;
fix: profile userData name and email won't show up;

tryproxy committed:
fix: redirect in burger constructor

tryproxy committed:
feat: pending order modal in burger contructor;
fix: header links have no correct styling;
fix: constructor in burger constructor won't remove items after creating order;
fix: selectors won\'t work - overload error

tryproxy committed:
fix: routing, modal popup;

tryproxy committed:
refactor: refactor selectors, clean up code
fix: ingredient count in burger constructor;

tryproxy committed:
build: add env.example

gennady-bars commented on src/utils/burger-api.ts:
```diff
@@ -75,7 +75,9 @@ export const getIngredientsApi = () =>
     .then((res) => checkResponse<TIngredientsResponse>(res))
     .then((data) => {
-      if (data?.success) return data.data;
+      if (data?.success) {
```
**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

gennady-bars commented on src/services/slices/profileSlice.ts:
```diff
@@ -0,0 +1,193 @@
+    password: string;
+    token: string;
+  }): Promise<{ success: boolean }> => resetPasswordApi({ password, token })
+);
```
**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде


```ts



export const fetchRegisterUser = createAsyncThunk<TAuthResponse, TRegisterData>(
  'profile/fetchRegisterUser',
  registerUserApi
);

export const fetchUser = createAsyncThunk<TUserResponse, undefined>(
  'profile/fetchUser',
  getUserApi
);

export const loginUser = createAsyncThunk<TAuthResponse, TLoginData>(
  'profile/userLogin',
  loginUserApi
);

export const logoutUser = createAsyncThunk<TServerResponse<{}>, undefined>(
  'profile/userLogout',
  logoutApi
);

export const userUpdate = createAsyncThunk<
  TUserResponse,
  Partial<TRegisterData>
>('profile/userUpdate', updateUserApi);

export const recoverPassUser = createAsyncThunk<
  { success: boolean },
  { email: string }
>('profile/userRecoverPas', forgotPasswordApi);

export const resetPassUser = createAsyncThunk<
  { success: boolean },
  { password: string; token: string }
>('profile/userResetPass', resetPasswordApi);

```

gennady-bars commented on src/services/slices/burgerConstructorSlice.ts:
```diff
@@ -0,0 +1,80 @@
+            sliceState.constructorItems.ingredients.push(action.payload);
+        }
+      },
+      prepare: (ingredient) => ({ payload: { ...ingredient, id: nanoid() } })
```
Отлично, что в `prepare` создаете `id`

gennady-bars commented on src/pages/login/login.tsx:
```diff
@@ -1,14 +1,25 @@
+import { useNavigate } from 'react-router-dom';
 
 export const Login: FC = () => {
   const [email, setEmail] = useState('');
   const [password, setPassword] = useState('');
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/pages/constructor-page/constructor-page.tsx:
```diff
@@ -1,15 +1,23 @@
+
+  const dispatch = useAppDispatch();
+  useEffect(() => {
+    dispatch(fetchIngredients());
```
`dispatch(fetchIngredients())`  достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

gennady-bars commented on src/components/protected-route/protected-route.tsx:
```diff
@@ -0,0 +1,26 @@
+  const location = useLocation();
+  const user = useAppSelector(userDataSelector);
+
+  if (!unAuthed && !(user.name && user.email)) {
+    return <Navigate to={`/login`} state={{ from: location }} replace />;
+  }
+
+  if (unAuthed && user.name && user.email) {
+    const fromPage = location.state?.from || { pathname: `/` };
+    return <Navigate to={fromPage} replace />;
+  }
```
Хорошо сделан `ProtectedRoute`

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,132 @@
+  const dispatch = useAppDispatch();
+  const navigate = useNavigate();
+  const location = useLocation();
+  const backgroundLocation = location.state?.background;
```
Отлично сделана логика отображения попапа и страницы  под ним

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,14 +1,132 @@
+          <Route
+            path='/ingredients/:id'
+            element={
+              <ProtectedRoute>
```
`<ProtectedRoute>` тут не нужно указывать

* Если я неавторизован, то при клике на ингредиенты меня перекидывает на логин, а защищать нужно только **профиль** и историю заказов внутри него

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что в `prepare` создаете `id`
* Хорошо сделан `ProtectedRoute`
* Отлично сделана логика отображения попапа и страницы  под ним

но есть некоторые недочеты:



* Если я неавторизован, то при клике на ингредиенты меня перекидывает на логин, а защищать нужно только **профиль** и историю заказов внутри него
* В попапе с деталями **заказа** не нужно указывать **Детали ингредиента** https://disk.yandex.ru/i/UGuqRG-JeumsJw 
* При клике на **старые** заказы из истории не отображаются детали этого заказа. Вам нужно делать запрос на сервер по номеру заказа, чтобы получить его данные. Не нужно искать их в заказах **ленты** (там нет старых заказов)
* `dispatch(fetchIngredients())`  достаточно 1 раз вызвать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.


**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде

tryproxy committed:
fix: title in order profile popup incorrect;
fix: show order-info for old orders in profile history;

tryproxy committed:
fix: remove redundant dispatch call;
fix: remove unnecessary protectedRout HOC for ingredients;
refactor: remove anonymous arrow fn in AsyncThunk

tryproxy committed:
feat: add matcher utility fn for extraReducers;
fix: error 401 in console caused by fetchUser;
feat: add custom hook to handle inputs in forms login, reset/forgot pass and register;

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -28,7 +27,6 @@ const App = () => {
   const backgroundLocation = location.state?.background;
 
   useEffect(() => {
-    dispatch(fetchUser());
```
`dispatch(fetchUser());` не нужно было удалять из `App`. Как раз это нужно для **автологина** (он пропал теперь после перезагрузки страницы)

gennady-bars reviewed (commented):
Уже почти все отлично, но:

* `dispatch(fetchUser());` не нужно было удалять из `App`. Как раз это нужно для **автологина** (он пропал теперь после перезагрузки страницы)

tryproxy committed:
fix: add fetchUser to app for autologin;

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
